sharing/handler.go

```diff
@@ -334,3 +390,139 @@ func makeLocation(base *url.URL, reqPath string) string {
 	location.Path = path.Join(location.Path, reqPath)
 	return location.String()
 }
+
+// handleHostingService deals with linksharing via custom URLs.
+func (handler *Handler) handleHostingService(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	host, _, err := net.SplitHostPort(r.Host)
+	if err != nil && strings.Contains(err.Error(), "missing port in address") {
+		host = r.Host
+	} else if err != nil {
+		handler.log.Error("unable to handle request", zap.Error(err))
+		http.Error(w, "unable to handle request", http.StatusInternalServerError)
+		return err
+	}
+
+	access, root, err := handler.getRootAndAccess(host)
+	if err != nil {
+		handler.log.Error("unable to handle request", zap.Error(err))
+		http.Error(w, "unable to handle request", http.StatusInternalServerError)
+		return err
+	}
+
+	project, err := uplink.OpenProject(ctx, access)
+	if err != nil {
+		handler.handleUplinkErr(w, "open project", err)
+		return err
+	}
+	defer func() {
+		if err := project.Close(); err != nil {
+			handler.log.With(zap.Error(err)).Warn("unable to close project")
+		}
+	}()
+
+	// e.g. http://mydomain.com/folder2/index.html with root="bucket1/folder1"
+	rootPath := strings.SplitN(root, "/", 2) // e.g. rootPath=[bucket1, folder1]
+	bucket := rootPath[0]
+	pt := strings.TrimPrefix(r.URL.Path, "/") // e.g. path = "folder2/index.html
+	if len(rootPath) == 2 {
+		pathPrefix := rootPath[1]  // e.g. pathPrefix = "folder1"
+		pt = pathPrefix + "/" + pt // e.g. path="folder1/folder2/index.html"
+	}
+
+	if pt == "" {
+		pt = "index.html"
+	}
+
+	o, err := project.StatObject(ctx, bucket, pt)
+	if err != nil {
+		handler.handleUplinkErr(w, "stat object", err)
+		return err
+	}
+
+	httpranger.ServeContent(ctx, w, r, pt, o.System.Created, objectranger.New(project, o, bucket))
+	return nil
+}
+
+// getRootAndAccess fetches the root and access grant from the cache or dns server when applicable.
+func (handler *Handler) getRootAndAccess(hostname string) (access *uplink.Access, root string, err error) {
+	record, exists := handler.checkCache(hostname)
+	if exists {
+		return record.access, record.root, nil
+	}
+	access, root, err = getRemoteRecord(hostname)
+	if err != nil {
+		return access, root, err
+	}
+	handler.updateCache(hostname, root, access)
+
+	return access, root, err
+}
+
+// checkCache checks the txt record cache to see if we have a valid access grant and root path.
+func (handler *Handler) checkCache(hostname string) (record txtRecord, exists bool) {
```

Comment:
is there a reason this isn't a method on txt records?
and the txtrecords in a different file?